benzina/src/array.rs

```diff
@@ -0,0 +1,269 @@
+use std::fmt::Debug;
+
+use diesel::{
+    deserialize::{FromSql, FromSqlRow},
+    expression::{AppearsOnTable, Expression, SelectableExpression},
+    pg::{Pg, PgValue},
+    query_builder::{AstPass, QueryFragment, QueryId},
+    result::QueryResult,
+    serialize::ToSql,
+    sql_types::{self, BigInt, Bool, Double, Float, Integer, Nullable, SmallInt, Text},
+};
+
+use crate::{U15, U31, U63, error::InvalidArray};
+
+/// A diesel [`Array`] serialization and deserialization wrapper __without__ NULL items
+///
+/// Since postgres ignores the array dimension (if specified), diesel implements
+/// [`FromSql`] for `Vec<T>` and [`ToSql`] for `Vec<T>`/`&[T]`.
+/// In addition postgres also considers array items as always nullable.
+/// This makes it hard to deal with real arrays that have a predetermined length
+/// and an homogeneous nullability.
+/// This type checks at runtime the array length and the __non__ nullability of its items,
+/// therefore to be completely safe, you should also add the following `CHECK` constraints:
+/// ```sql
+/// array_ndims(array_field) = 1 AND
+/// array_length(array_field, 1) = N AND
+/// array_position(array_field, NULL) IS NULL
+/// ```
+///
+/// This type is not intended to be used directly in the model but rather to be
+/// used with diesel [`serialize_as`] and [`deserialize_as`].
+///
+/// ```
+/// use benzina::{Array, U31};
+/// use diesel::{Insertable, Queryable};
+///
+/// #[derive(Debug, Queryable)]
+/// #[diesel(table_name = users, check_for_backend(diesel::pg::Pg))]
+/// struct User {
+///     id: U31,
+///     first_name: String,
+///     last_name: String,
+///     #[diesel(deserialize_as = Array<bool, 5>)]
+///     flags: UserFlags,
+/// }
+///
+/// #[derive(Debug, Insertable)]
+/// #[diesel(table_name = users)]
+/// struct NewUser {
+///     first_name: String,
+///     last_name: String,
+///     #[diesel(serialize_as = Array<bool, 5>)]
+///     flags: UserFlags,
+/// }
+///
+/// #[derive(Debug)]
+/// struct UserFlags([bool; 5]);
+///
+/// // needed by deserialize_as
+/// impl From<benzina::Array<bool, 5>> for UserFlags  {
+///    fn from(value: benzina::Array<bool, 5>) -> Self {
+///        Self(value.into_inner())
+///    }
+/// }
+///
+/// // needed by serialize_as
+/// impl From<UserFlags> for benzina::Array<bool, 5> {
+///    fn from(value: UserFlags) -> Self {
+///        Self::new(value.0)
+///    }
+/// }
+///
+/// diesel::table! {
+///     users (id) {
+///         id -> Int4,
+///         first_name -> Text,
+///         last_name -> Text,
+///         flags -> Array<Nullable<Bool>>,
+///     }
+/// }
+/// ```
+///
+/// [`Array`]: diesel::sql_types::Array
+/// [`serialize_as`]: diesel::prelude::Insertable#optional-field-attributes
+/// [`deserialize_as`]: diesel::prelude::Queryable#deserialize_as-attribute
+#[derive(Debug, FromSqlRow)]
+pub struct Array<T, const N: usize>([T; N]);
+impl<T, const N: usize> Array<T, N> {
```

Comment:
This may make it easier to read, since the array is externally visible within the `Array`.

```suggestion
pub struct Array<T>(T);
impl<T, const N: usize> Array<[T; N]> {
```